### DISCUSSION TITLE
Filter vs shortcode escaping mechanism

### DISCUSSION THREAD

**Discussion by pavelloz** [11ty/eleventy#2256] [Q&A] [answered]
Hello,

I tried to make a filter, but it escapes quotes (at least) and i could not find a way to unescape them. Fortunately shortcodes do not do this, so im using shortcode, but i was wondering if there is a way of forcing filter to not escape anything:
  
Filter/shortcode:
```js
    const prod = process.env.NODE_ENV === 'production';

  eleventyConfig.addFilter("asset_url", (path) => {
    return prod ? `{{ '${path}' | asset_url }}` : `/assets/${path}`;
  });

  eleventyConfig.addShortcode("asset_url", function(path) {
    return prod ? `{{ '${path}' | asset_url }}` : `/assets/${path}`;
  });
```

Input njk:
```html
    <link rel="stylesheet" type="text/css" href="{% asset_url 'main.css' %}" />
    <link rel="stylesheet" type="text/css" href="{{ 'main.css' | asset_url }}" />
```

Output html:
```html
    <link rel="stylesheet" type="text/css" href="{{ 'main.css' | asset_url }}" />
    <link rel="stylesheet" type="text/css" href="{{ &#39;main.css&#39; | asset_url }}" />
```

I tried adding `| safe` and various JS functions in JS code to unescape, but nothing worked so far.

Thanks in advance for any input :)

**Comment by pdehaan** [comment]
Ah, I was going to suggest the `safe` Filter, but I believe your issue is that your nunjucks filter/shortcode returns nunjucks output which isn’t reparsed by nunjucks. Even if nunjucks parsed the files twice, your filter version would be an infinite loop since it references itself if you’re doing a production build.

maybe something like this?
```js
  eleventyConfig.addFilter("asset_url", function (path) {
    const prod = process.env.NODE_ENV === 'production';
    return prod ? path : `/assets/${path}`;
  });
```

**Reply by pavelloz** [reply]
I kind of need it to be `{{ 'x' | asset_url }}` at the end of the day, because output needs to be a liquid syntax that will be later on interpeted on the server as well.

If i understand your solution, i would need to pass the whole {{ 'x' | asset_url }} into the filter, which would still escape quotes.

So far im liking shortcodes (amazing feature), so its not a big deal, but to keep parity (i want to make users not worry about any migrations) with liquid syntax i was just wondering if escaping by default just can be switched off, or at least mitigated.

**Comment by pavelloz** [comment, marked as answer]
Found it!

I switched all nunjucks templates to liquid, so i used addLiquidFilter instead of generic one, and it works now.

```js
  eleventyConfig.addLiquidFilter("asset_url", function (path) {
    return prod ? `{{ '${path}' | asset_url }}` : `/assets/${path}`;
  });
```

Thank you for inspiration and your help :)
